crates/engine/local/src/service.rs

```diff
@@ -0,0 +1,298 @@
+
+        Ok(new_head.hash)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use reth_chainspec::MAINNET;
+    use reth_config::PruneConfig;
+    use reth_db::test_utils::{create_test_rw_db, create_test_static_files_dir};
+    use reth_ethereum_engine_primitives::EthEngineTypes;
+    use reth_exex_test_utils::TestNode;
+    use reth_node_types::NodeTypesWithDBAdapter;
+    use reth_payload_builder::test_utils::spawn_test_payload_service;
+    use reth_primitives::B256;
+    use reth_provider::{providers::StaticFileProvider, BlockReader, ProviderFactory};
+    use reth_prune::PrunerBuilder;
+    use reth_transaction_pool::{
+        test_utils::{testing_pool, MockTransaction},
+        TransactionPool,
+    };
+    use std::{convert::Infallible, time::Duration};
+    use tokio::sync::mpsc::unbounded_channel;
+
+    #[derive(Debug)]
+    struct TestPayloadAttributesBuilder;
+
+    impl PayloadAttributesBuilder for TestPayloadAttributesBuilder {
+        type PayloadAttributes = reth_rpc_types::engine::PayloadAttributes;
+        type Error = Infallible;
+
+        fn build(&self) -> Result<Self::PayloadAttributes, Self::Error> {
+            Ok(reth_rpc_types::engine::PayloadAttributes {
+                timestamp: 0,
+                prev_randao: Default::default(),
+                suggested_fee_recipient: Default::default(),
+                withdrawals: None,
+                parent_beacon_block_root: None,
+            })
+        }
+    }
+
+    #[tokio::test]
+    async fn test_local_engine_service_interval() -> eyre::Result<()> {
+        reth_tracing::init_test_tracing();
+
+        // Start the provider and the pruner
+        let (_, static_dir_path) = create_test_static_files_dir();
+        let provider = ProviderFactory::<NodeTypesWithDBAdapter<TestNode, _>>::new(
+            create_test_rw_db(),
+            MAINNET.clone(),
+            StaticFileProvider::read_write(static_dir_path).unwrap(),
+        );
+        let pruner = PrunerBuilder::new(PruneConfig::default())
+            .build_with_provider_factory(provider.clone());
+
+        // Start the payload builder service
+        let payload_handle = spawn_test_payload_service::<EthEngineTypes>();
+
+        // Sync metric channel
+        let (sync_metrics_tx, _) = unbounded_channel();
+
+        // Get the attributes for start of block building
+        let genesis_hash = B256::random();
+
+        // Launch the LocalEngineService in interval mode
+        let period = Duration::from_secs(1);
+        LocalEngineService::spawn_new(
+            payload_handle,
+            TestPayloadAttributesBuilder,
+            provider.clone(),
+            pruner,
+            sync_metrics_tx,
+            genesis_hash,
+            MiningMode::interval(period),
+        );
+
+        // Wait 4 intervals
+        tokio::time::sleep(4 * period).await;
+
+        // Assert a block has been build
+        let block = provider.block_by_number(0)?;
```

Comment:
this doesn't really work yet — if you change it to `block_by_number(1)` it would fail because currently it uses `TestPayloadJobGenerator` which doesn't allow us to produce valid blocks

@mattsse do we have examples of such engine testing, I believe it should be straightforward to construct a working payload builder, but I'm wondering if such tests belong somewhere else?

Reply:
I guess the idea of the test here was just to check that a block was built, not a valid block. The second test showcases this a little better, as it checks that `block_by_number(0)` returns `None`, adds a transaction to the pool and verifies that this triggers block building.

I can for sure update the test if you have test utils I can use to build valid blocks.